torch/_inductor/comms.py

```diff
@@ -17,61 +20,116 @@
-    for snode in reversed(snodes):
-        if is_collective(snode.node):
-            cur_comms.append(snode)
-        else:
-            for comm in cur_comms:
-                assert len(inverse_users[comm]) > 0
-            while len(cur_comms) > 0 and any(
-                snode in inverse_users[comm] for comm in cur_comms
-            ):
-                comm = cur_comms.pop(0)
-                new_order_reversed.append(comm)
-            new_order_reversed.append(snode)
-    assert len(cur_comms) <= 1
-    new_order_reversed.extend(tuple_sorted(cur_comms))
-    return new_order_reversed[::-1]
+    return raise_comms_and_sink_waits(snodes, raise_comms=True, sink_waits=False)
+
+
+def raise_comms_and_sink_waits(
+    snodes: List[BaseSchedulerNode],
+    *args,
+    raise_comms: bool = True,
+    sink_waits: bool = True,
+) -> List[BaseSchedulerNode]:
+    # We assign each node a tuple of scores (score_0, score_1, score_2),
+    # decreasing in importance, with a lower value indicating a higher ranking:
+    #
+    # - score_0: the lowest comm_idx among the comm nodes that the node blocks.
+    # If a node doesn't block any comm nodes, its score_0 is set to
+    # sys.maxsize. This score ensures that comm nodes get scheduled as early as
+    # possible.
+    # - score_1: 1 if the node is a wait node, 0 otherwise. This score ensures
+    # that wait nodes are deferred as late as possible.
+    # - score_2: the index of the node in the original topological order. This
+    # score provides stability in case of ties.
+    #
+    # When only raise_comms is True, only score_0 and score_2 are considered.
+    # When only sink_waits is True, only score_1 and score_2 are considered.
+    # When neither is True, the original order is yielded.
+    name_to_snode = {}
+    scores_0, scores_1, scores_2 = {}, {}, {}
+    for idx, snode in enumerate(snodes):
+        for name in snode.get_names():
+            name_to_snode[name] = snode
+            scores_0[name] = sys.maxsize
+            scores_1[name] = 0
+            scores_2[name] = idx
+
+    comm_idx = 0
+    for snode in snodes:
+        if raise_comms and is_collective(snode.node):
+            scores_0[snode.get_name()] = comm_idx
+            for anc in snode.ancestors:
+                scores_0[anc] = min(scores_0[anc], comm_idx)
+            comm_idx += 1
+        elif sink_waits and is_wait(snode.node):
+            scores_1[snode.get_name()] = 1
+
+    class Runnable:
+        def __init__(self, snode):
+            self.snode = snode
+            name = next(iter(snode.get_names()))
+            self.score = (
+                scores_0[name],
+                scores_1[name],
+                scores_2[name],
+            )
+
+        def __lt__(self, other):
+            return self.score < other.score
+
+    unmet_deps: Dict[BaseSchedulerNode, Set[str]] = {}
+    for snode in snodes:
+        # A mutating node's unmet_dependencies doesn't cover the dependencies
```

Comment:
I don't think this should be needed - the mutation dependencies should already be encoded in the scheduler node's edges.

Reply:
I thought so too but it doesn't seem to be the case (learned it the hard way). For example, we have:
```
node = (...)
mutation_output = (node_doing_mutating=node, ...)
```

`mutation_output` would have more `unmet_dependencies` than `node`.

Maybe we should fix it in the scheduler?

Reply:
Yes `MutationOutput` expressed the dependencies incorrectly. This is fixed in #129346 though (the entire stack).

Reply:
@peterbell10 appreciate the fix!

Reply:
after the rebase, we will need some changes (https://github.com/pytorch/pytorch/pull/130276) to make this algo working again (please feel free to merge my PR into this PR 🙂)